### PR TITLE
More optimizations for Shared Component Properties

### PR DESCRIPTION
**Pull request by adamcin**
This patch includes changes to leverage standard Sling object patterns to greatly improve performance of the SharedValueMapValueInjector and the SharedComponentPropertiesBindingsValuesProvider when rendering a complex page.

* SharedComponentPropertiesBindingsValuesProvider uses an internal request attribute cache type for collecting Bindings computed different levels of shared properties according to their resource paths. This optimizes the BVP behavior by keeping the SCP results around for reuse across sling includes.
* The SharedValueMapValueInjector will now attempt to retrieve results from the SlingBindings when the model adaptable is a request. This greatly optimizes requests for the simple case where a sling model used within an HTL script can access any SCP-provided properties from the SlingBindings, rather having to evaluate the adaptable request or resource again. 
* For more complicated scenarios where the injector is evaluating SCP for a resource adaptable, such as when it is referenced by another model using the @ChildResource annotation, and the bindings provided by BVP are not present or are not relevant for the child resource, the injector will adapt the resource to a SharedValueMapResourceAdapter, which implicitly leverages the SlingAdaptable adapters cache so that the SCP calculation for the injector is only evaluated once. This is similar to the traditional behavior of Resource.getValueMap() and originally Resource.adaptTo(ValueMap.class).

### CODE REVIEW

File: bundle/src/main/java/com/adobe/acs/commons/models/injectors/impl/SharedValueMapValueInjector.java [hunk 1]
```diff
@@ -21,106 +21,234 @@
...
+            } else { // otherwise, pass null for sharedProperties and merge with global and resource
+                return sharedComponentProperties.mergeProperties(globalVmBound, null, resource);
+            }
+        }
+
+        // it is also possible for the shared properties path to differ from the path in bindings for the same
+        // reason that shared properties page path may differ (a wrapped request for a child resource of a
+        // different resource type)
+        final ValueMap sharedVmBound;
+        if (sharedPropertiesPath.equals(bindings.get(SharedComponentProperties.SHARED_PROPERTIES_PATH))) {
```
Comment (HitmanInWis) [inline bundle/src/main/java/com/adobe/acs/commons/models/injectors/impl/SharedValueMapValueInjector.java:196]: Would it have been a simpler system to use a property like "shared-propertied-<path>" so that we would only need a single property instead of both `SHARED_PROPERTIES_PATH` and `SHARED_PROPERTIES` and then technically could have bindings for multiple resources?

**Review by HitmanInWis — CHANGES_REQUESTED** [review]

File: bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedPropertiesRequestCache.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import javax.script.SimpleBindings;
+import javax.servlet.ServletRequest;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Consumer;
+
+/**
+ * Simple cache for global and shared properties bindings keyed by path and persisted in a request attribute.
+ */
+final class SharedPropertiesRequestCache {
```
Comment (HitmanInWis) [inline bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedPropertiesRequestCache.java:33]: Is this class purposely package protected?

File: bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedPropertiesRequestCache.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            cache = new SharedPropertiesRequestCache();
+            cache.toRequest(req);
+        }
+        return cache;
+    }
+
+    public SharedPropertiesRequestCache toRequest(ServletRequest req) {
+        SharedPropertiesRequestCache prev = (SharedPropertiesRequestCache) req.getAttribute(REQUEST_ATTRIBUTE_NAME);
+        req.setAttribute(REQUEST_ATTRIBUTE_NAME, this);
+        return prev;
```
Comment (HitmanInWis) [inline bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedPropertiesRequestCache.java:66]: What is the purpose of returning `prev` here?

File: bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedValueMapResourceAdapterImpl.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
+/*
+ * #%L
+ * ACS AEM Commons Bundle
+ * %%
+ * Copyright (C) 2016 Adobe
```
Comment (HitmanInWis) [inline bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedValueMapResourceAdapterImpl.java:5]: nit: copyrights on new files should have current year

**Review by HitmanInWis — CHANGES_REQUESTED** [review]

File: bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedComponentPropertiesBindingsValuesProvider.java [hunk 5]
```diff
@@ -57,107 +55,81 @@
...
     private static final Logger log = LoggerFactory.getLogger(SharedComponentPropertiesBindingsValuesProvider.class);
 
-    @Reference
-    private PageRootProvider pageRootProvider;
-
-    @Reference
+    /**
+     * Bind if available, check for null when reading.
+     */
+    @Reference(policyOption = ReferencePolicyOption.GREEDY, cardinality = ReferenceCardinality.OPTIONAL_UNARY)
```
Comment (HitmanInWis) [inline bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedComponentPropertiesBindingsValuesProvider.java:61]: Was this a change so that using a SCP annotation on a model w/o SCP configured wont blow things up?

File: bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedComponentPropertiesBindingsValuesProvider.java [hunk 6]
```diff
@@ -57,107 +55,81 @@
...
+    public void addBindings(final Bindings bindings) {
+        final SlingHttpServletRequest request = (SlingHttpServletRequest) bindings.get(SlingBindings.REQUEST);
+        final Resource resource = (Resource) bindings.get(SlingBindings.RESOURCE);
+        if (request != null && resource != null) {
+            final SharedPropertiesRequestCache cache = SharedPropertiesRequestCache.fromRequest(request);
+            if (sharedComponentProperties != null) {
+                setSharedProperties(bindings, resource, cache);
             } else {
-                log.debug("Page Root Provider must be configured for shared component properties to be supported");
+                log.debug("Shared Component Properties must be configured enable this provider");
```
Comment (HitmanInWis) [inline bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedComponentPropertiesBindingsValuesProvider.java:73]: nit: "to" enable this provider

File: bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedComponentPropertiesBindingsValuesProvider.java [hunk 7]
```diff
@@ -57,107 +55,81 @@
...
     @Override
-    public void addBindings(Bindings bindings) {
-        Resource resource = (Resource) bindings.get("resource");
-        if (resource != null) {
-            if (pageRootProvider != null) {
-                setSharedProperties(bindings, resource);
+    public void addBindings(final Bindings bindings) {
+        final SlingHttpServletRequest request = (SlingHttpServletRequest) bindings.get(SlingBindings.REQUEST);
+        final Resource resource = (Resource) bindings.get(SlingBindings.RESOURCE);
+        if (request != null && resource != null) {
```
Comment (HitmanInWis) [inline bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedComponentPropertiesBindingsValuesProvider.java:68]: Previous code only checked for `resource != null` but now we're also requiring `request` to be non-null.  Does this add any risks?

**Review by HitmanInWis — CHANGES_REQUESTED** [review]

File: bundle/src/main/java/com/adobe/acs/commons/models/injectors/impl/SharedValueMapValueInjector.java [hunk 8]
```diff
@@ -21,106 +21,234 @@
...
     }
 
     /**
      * Get shared properties ValueMap the current resource.
      */
-    protected ValueMap getSharedProperties(String rootPagePath, Resource resource) {
-        String sharedPropsPath = rootPagePath + "/" + JcrConstants.JCR_CONTENT + "/" + SharedComponentProperties.NN_SHARED_COMPONENT_PROPERTIES + "/" + resource.getResourceType();
-        Resource sharedPropsResource = resource.getResourceResolver().getResource(sharedPropsPath);
-        return sharedPropsResource != null ? sharedPropsResource.getValueMap() : ValueMap.EMPTY;
+    protected ValueMap getSharedProperties(Resource resource) {
```
Comment (HitmanInWis) [inline bundle/src/main/java/com/adobe/acs/commons/models/injectors/impl/SharedValueMapValueInjector.java:227]: Should `getSharedProperties` and `getGlobalProperties` be consolidated in the new `SharedComponentPropertiesImpl` since we also need it there for `getSharedValueMapResourceAdapter`?

**Review by HitmanInWis — CHANGES_REQUESTED** [review]

File: bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedComponentPropertiesBindingsValuesProvider.java (hunk 7 above)
Comment (adamcin) [reply bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedComponentPropertiesBindingsValuesProvider.java:68]: No. A BindingsValuesProvider is only executed in the context of a Sling request, and should have access to all property names defined by the SlingBindings interface. But we still want to check for null, in case, you know, whatever.

File: bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedComponentPropertiesBindingsValuesProvider.java (hunk 5 above)
Comment (adamcin) [reply bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedComponentPropertiesBindingsValuesProvider.java:61]: Yes, exactly. I want the annotation behavior to "failover gracefully" to the default bindings, like I'm doing with @SharedValueMapValue behaving like @ValueMapValue.

File: bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedPropertiesRequestCache.java (hunk 3 above)
Comment (adamcin) [reply bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedPropertiesRequestCache.java:66]: Just followed the pattern of WCMMode.toRequest(). Didn't have any other reason.

File: bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedPropertiesRequestCache.java (hunk 2 above)
Comment (adamcin) [reply bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedPropertiesRequestCache.java:33]: No. I think the IDE does that when I move an inner class up a level. I'll make it public.

File: bundle/src/main/java/com/adobe/acs/commons/models/injectors/impl/SharedValueMapValueInjector.java [hunk 9]
```diff
@@ -21,106 +21,234 @@
...
+            case MERGED:
+                return Optional.ofNullable(adapter)
+                        .map(SharedValueMapResourceAdapter::getMergedProperties)
+                        .orElseGet(() -> getMergedProperties(resource));
+            default:
+                return ValueMap.EMPTY;
+        }
+    }
+
+    ValueMap getValueMapFromBindings(final SlingBindings bindings,
```
[the remark this answers is not in the document]
Comment (adamcin) [reply bundle/src/main/java/com/adobe/acs/commons/models/injectors/impl/SharedValueMapValueInjector.java:154]: I'll correct that.

File: bundle/src/main/java/com/adobe/acs/commons/models/injectors/impl/SharedValueMapValueInjector.java (hunk 1 above)
Comment (adamcin) [reply bundle/src/main/java/com/adobe/acs/commons/models/injectors/impl/SharedValueMapValueInjector.java:196]: I suppose that would work for how we're using the bindings with the SCP API, but it would be a bit unexpected for implicit script variable names, which are taken from the binding property name. I assume they should be legal java identifiers. I'm not sure what would happen with hyphens or slashes.

File: bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedValueMapResourceAdapterImpl.java (hunk 4 above)
Comment (adamcin) [reply bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedValueMapResourceAdapterImpl.java:5]: they're not copyrights. they're copypastes.

File: bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedComponentPropertiesBindingsValuesProvider.java (hunk 6 above)
Comment (adamcin) [reply bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedComponentPropertiesBindingsValuesProvider.java:73]: will correct.

File: bundle/src/main/java/com/adobe/acs/commons/models/injectors/impl/SharedValueMapValueInjector.java (hunk 8 above)
Comment (adamcin) [reply bundle/src/main/java/com/adobe/acs/commons/models/injectors/impl/SharedValueMapValueInjector.java]: done.

File: bundle/src/main/java/com/adobe/acs/commons/models/injectors/impl/SharedValueMapValueInjector.java (hunk 1 above)
Comment (HitmanInWis) [reply bundle/src/main/java/com/adobe/acs/commons/models/injectors/impl/SharedValueMapValueInjector.java:196]: Ok, my only reason for thinking this was in case it somehow enabled better caching of bindings for all the things w/o overwriting each other back and forth for alternating components of different resource types.  I haven't yet run the code, so I dont know if that's applicable or not.  I just see code validating that current bindings are for the current resource, which made me think that it might be possible that bindings are flip flopping, which could be prevented via naming.

Perhaps another way we could resolve would be to put the bindings into a map object, keyed by resource path.

Comment (adamcin) [reply bundle/src/main/java/com/adobe/acs/commons/models/injectors/impl/SharedValueMapValueInjector.java:196]: It's not so much that they are flip-flopping, because if that happens the caching is still being managed by BVP at the start of the servlet execution for a requested resource. The check for the path is to protect against the case where the cached bindings AREN'T overwritten, when one might assume they should be. I.E. when a Request-adaptable model has a @ChildResourceWithRequest-annotated model member that has its own @SharedValueMapValue annotations. The BVP doesn't have a chance to execute for the request wrapper we inject for the child model, but the parent bindings are still present because we carry along the SlingBindings request attribute in the wrapper class. This means that if we naively check for the presence of a "sharedProperties" binding, it will be for the parent resource, not the current resource. We haven't actually created a new sharedProperties for the child because we are skipping the normal BindingsValuesProvider phase with the wrapped request, so there is no relevant cache value to look up.

Comment (HitmanInWis) [reply bundle/src/main/java/com/adobe/acs/commons/models/injectors/impl/SharedValueMapValueInjector.java:196]: In that situation, does the changes to the bind values from the child object affect the original request pointing at the parent object?

Comment (adamcin) [reply bundle/src/main/java/com/adobe/acs/commons/models/injectors/impl/SharedValueMapValueInjector.java:196]: I'm saying those binding values don't change at all for the child resource. The Injector operating on the child doesn't push any changes back to the request bindings. It only decides whether or not to use the bindings based on whether the path bindings match the values that would be expected for child resource. If not, it ignores the bindings altogether.

Comment (HitmanInWis) [reply bundle/src/main/java/com/adobe/acs/commons/models/injectors/impl/SharedValueMapValueInjector.java:196]: ah k, got it

**Comment by HitmanInWis** [comment]
Code all looks good to me - I just need to find time to test.

**Comment by HitmanInWis** [comment]
Currently testing, hoping to get thru it today :)

**Comment by HitmanInWis** [comment]
Ok, I've tested this very thoroughly thru all the different code branches on AEM 6.5.  Unfortunately won't be able to get Cloud Service and 6.4 until early next week, but feeling pretty good about these optimizations being backward compatible and safe.

File: bundle/src/main/java/com/adobe/acs/commons/models/injectors/impl/SharedValueMapValueInjector.java [hunk 10]
```diff
@@ -21,106 +21,234 @@
...
+        // we always need a resource, if only to determine if cached global properties are valid for the resource path
+        if (resource == null) {
+            return ValueMap.EMPTY;
+        }
+
+        // the root page path is used to test the validity of bindings
+        final String rootPagePath = sharedComponentProperties.getSharedPropertiesPagePath(resource);
+        if (rootPagePath == null) {
+            // when we have a resource but no root page path, we can at least satisfy MERGED with the resource valuemap
+            return valueType == SharedComponentProperties.ValueTypes.MERGED ? resource.getValueMap() : null;
```
Comment (HitmanInWis) [inline bundle/src/main/java/com/adobe/acs/commons/models/injectors/impl/SharedValueMapValueInjector.java:121]: edge case bug here, where SCP is configured but there is no page root - the `else` case returns `null` but needs to instead return `ValueMap.EMPTY` if we dont want model objects pulling `GLOBAL` or `SHARED` `valueType` to blow up.

**Review by HitmanInWis — CHANGES_REQUESTED** [review]

**Review by HitmanInWis — APPROVED** [review]

**Comment by adamcin** [comment]
@davidjgonzalez I think we need to upgrade the codecov upload step in the github workflow. My fa2f2a2 commit above is unfortunately not going to affect the active workflow defined in the master branch until it is merged.

**Comment by davidjgonzalez** [comment]
yeah - the github actions broke the ability to send to Codecov on PRs - was looking into it a while ago (think theres an open PR i made to test it out) .. iIRC we might have to add some override vars to get it to work? Or maybe Codecov has made updates on their side since then.

**Comment by kwin** [comment]
> @davidjgonzalez any idea why the `codecov/project` test never comes back? Looks like it's preventing me from merging this for Mark.

Not sure, seems to be an edge case with PRs. We are trying out a new GH action in https://github.com/Adobe-Consulting-Services/acs-aem-commons/pull/2706/files. Feel free to comment. For the time being I no longer made that status required.

File: .github/workflows/maven.yml [hunk 11]
```diff
@@ -93,12 +93,14 @@ jobs:
 
       - name: Upload code coverage to CodeCov (Main build)
         if: matrix.isMainBuildEnv
-        run: bash <(curl -s https://codecov.io/bash)
+        uses: codecov/codecov-action@v2
```
Comment (kwin) [inline .github/workflows/maven.yml:96]: @adamcin Please revert this commit. We have https://github.com/Adobe-Consulting-Services/acs-aem-commons/pull/2706/files for that.

Comment (adamcin) [reply .github/workflows/maven.yml]: @kwin reverted
